packages/flutter/lib/src/widgets/_window.dart

```diff
@@ -0,0 +1,553 @@
+  void onWindowCloseRequested(RegularWindowController controller) {
+    if (!isWindowingEnabled) {
+      throw UnsupportedError(_windowingDisabledErrorMessage);
+    }
+
+    controller.destroy();
+  }
+
+  /// Invoked when the window is closed.
+  ///
+  /// {@macro flutter.widgets.windowing.experimental}
+  @internal
+  void onWindowDestroyed() {
+    if (!isWindowingEnabled) {
+      throw UnsupportedError(_windowingDisabledErrorMessage);
+    }
+
+    final WindowingOwner owner = WidgetsBinding.instance.windowingOwner;
+    if (!owner.hasTopLevelWindows()) {
+      // TODO(mattkae): close the application if this is the last window
+      // via ServicesBinding.instance.exitApplication(AppExitType.cancelable);
+    }
+  }
+}
+
+/// A controller for a regular window.
+///
+/// A regular window is a traditional window that can be resized, minimized,
+/// maximized, and closed. Upon construction, the window is created for the
+/// platform with the provided properties.
+///
+/// This class does not interact with the widget tree. Instead, it is typically
+/// provided to the [RegularWindow] widget, who does the work of rendering the
+/// content inside of this window.
+///
+/// An example usage might look like:
+/// ```dart
+/// final RegularWindowController controller = RegularWindowController(
+///   contentSize: const WindowSizing(
+///     preferredSize: Size(800, 600),
+///     preferredConstraints: BoxConstraints(minWidth: 640, minHeight: 480),
+///   ),
+///   title: "Example Window",
+/// );
+/// runWidget(
+///   RegularWindow(
+///     controller: controller,
+///     child: MaterialApp(home: Container()),
+///   ),
+/// );
+/// ```
+///
+/// When provided to a [RegularWindow] widget, widgets inside of the [child]
+/// parameter will have access to the [RegularWindowController] via the
+/// [WindowControllerContext] widget.
+///
+/// {@macro flutter.widgets.windowing.experimental}
+@internal
+abstract class RegularWindowController extends BaseWindowController {
+  /// Creates a [RegularWindowController] with the provided properties.
+  ///
+  /// Upon construction, the window is created for the platform.
+  ///
+  /// The [preferredContentSize] argument sets the window's initial size preference.
+  /// This might not be honored by the platform.
+  ///
+  /// The [title] argument configures the window's initial title.
+  /// If omitted, some platforms might fall back to the app's name.
+  ///
+  /// The [delegate] argument can be used to listen to the window's
+  /// lifecycle. For example, it can be used to save state before
+  /// a window is closed.
+  ///
+  /// {@macro flutter.widgets.windowing.experimental}
+  @internal
+  factory RegularWindowController({
+    required WindowSizing preferredContentSize,
+    String? title,
+    RegularWindowControllerDelegate? delegate,
+  }) {
+    if (!isWindowingEnabled) {
+      throw UnsupportedError(_windowingDisabledErrorMessage);
+    }
+
+    WidgetsFlutterBinding.ensureInitialized();
```

Comment:
This is outside of my area of expertise and I'll defer to @justinmc here, but I suspect we don't want to do this. I _believe_ it's better to error if the user creates a `RegularWindowController` without an initialized widgets binding. Tests might want `TestWidgetsFlutterBinding` instead. Or out-of-tree embedders might want their own widgets binding instead. This might cause an annoying situation where it's difficult to find out why the wrong widgets binding was registered.